### ISSUE TITLE
Proyecto: Clasificación: Optimización del Modelo

### ISSUE THREAD

**Issue by furrutiav** [furrutiav/lab-mds-2022#16]
*Aquí deberán mejorar del modelo de clasificación al variar los algoritmos/hiperparámetros que están ocupando.*

- Generar una nueva `Pipeline` enfocada en buscar el mejor modelo usando GridSearch.
- Usar **`GridSearchCV`** o **`HalvingGridSearchCV`** para tunear hipermarámetros. La primera demorará más que la segunda pero les traerá potencialmente mejores resultados. Pueden probar también [`OptunaSearchCV`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.OptunaSearchCV.html) de la librería [`Optuna`](https://optuna.org/) , la cuál es bastante popular para buscar modelos de redes neuronales.
- Agregar técnicas de seleccion de atributos, como también usar mejores clasificadores y explorar sus hiperparámetros. 
- Probar distintos parámetros para las transformaciones de datos, seleccion de atributos, clasificadores, etc...
- Probar modelos basados en gradient boosting/bagging. **Recomendación fuerte:** Probar [`LightGMB`](https://lightgbm.readthedocs.io/en/latest/) o [`xgboost`](https://xgboost.readthedocs.io/en/stable/).
- Probar activando/descativando los procesadores de texto, de categorías, etc...
- Recuerden setear la búsqueda para optimizar la métrica que se evalua en la competencia.

Algunas notas interesantes sobre este proceso: 

- No se les pide rendimientos cercanos al 100% de la métrica para concretar exitosamente el proyecto. Por otra parte, celebren cada progreso que obtengan.
- Hagan grillas computables: Si la grilla se va a demorar 1/3 la edad del universo en explorarse completamente, entonces achíquenla a algo que sepan que va a terminar. 
- Aprovechen el procesamiento paralelo (con `njobs`) para acelerar la búsqueda. Sin embargo, si tienen problemas con la memoria RAM, reduzca la cantidad de jobs a algo que su computador/interprete web pueda procesar.

**Al final de este proceso, seleccione el mejor modelo de clasificación encontrado, prediga las labels del test set de la competencia y envíelos a Codalab.**

**Comment by furrutiav** [comment]
{'memory': None,
 'steps': [('FunctionTransformer',
   FunctionTransformer(func=<function feature_extractor at 0x00000234FB76D160>)),
  ('ColumnTransformer',
   ColumnTransformer(transformers=[('OneHot',
                                    OneHotEncoder(handle_unknown='ignore'),
                                    ['original_language', 'clusters_keywords',
                                     'clusters_overview', 'clusters_tagline']),
                                   ('MinMax', MinMaxScaler(),
                                    ['budget', 'release_date_month',
                                     'release_date_day',
                                     'num_top_production_companies',
                                     'num_top_artists', 'ratio(runtime, budget)',
                                     'num_artists', 'num_genres',
                                     'num_product...
                                   ('OneHot_split_credits',
                                    CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x00000234940BA670>),
                                    'credits'),
                                   ('OneHot_split_production_companies',
                                    CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x0000023494360BE0>),
                                    'production_companies'),
                                   ('OneHot_split_keywords',
                                    CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x0000023494360040>),
                                    'keywords')])),
  ('selector', SelectPercentile(percentile=95)),
  ('clf', LinearSVC(C=0.001, class_weight='balanced', random_state=42))],
 'verbose': False,
 'FunctionTransformer': FunctionTransformer(func=<function feature_extractor at 0x00000234FB76D160>),
 'ColumnTransformer': ColumnTransformer(transformers=[('OneHot',
                                  OneHotEncoder(handle_unknown='ignore'),
                                  ['original_language', 'clusters_keywords',
                                   'clusters_overview', 'clusters_tagline']),
                                 ('MinMax', MinMaxScaler(),
                                  ['budget', 'release_date_month',
                                   'release_date_day',
                                   'num_top_production_companies',
                                   'num_top_artists', 'ratio(runtime, budget)',
                                   'num_artists', 'num_genres',
                                   'num_product...
                                 ('OneHot_split_credits',
                                  CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x00000234940BA670>),
                                  'credits'),
                                 ('OneHot_split_production_companies',
                                  CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x0000023494360BE0>),
                                  'production_companies'),
                                 ('OneHot_split_keywords',
                                  CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x0000023494360040>),
                                  'keywords')]),
 'selector': SelectPercentile(percentile=95),
 'clf': LinearSVC(C=0.001, class_weight='balanced', random_state=42),
 'FunctionTransformer__accept_sparse': False,
 'FunctionTransformer__check_inverse': True,
 'FunctionTransformer__func': <function __main__.feature_extractor(X)>,
 'FunctionTransformer__inv_kw_args': None,
 'FunctionTransformer__inverse_func': None,
 'FunctionTransformer__kw_args': None,
 'FunctionTransformer__validate': False,
 'ColumnTransformer__n_jobs': None,
 'ColumnTransformer__remainder': 'drop',
 'ColumnTransformer__sparse_threshold': 0.3,
 'ColumnTransformer__transformer_weights': None,
 'ColumnTransformer__transformers': [('OneHot',
   OneHotEncoder(handle_unknown='ignore'),
   ['original_language',
    'clusters_keywords',
    'clusters_overview',
    'clusters_tagline']),
  ('MinMax',
   MinMaxScaler(),
   ['budget',
    'release_date_month',
    'release_date_day',
    'num_top_production_companies',
    'num_top_artists',
    'ratio(runtime, budget)',
    'num_artists',
    'num_genres',
    'num_production_companies',
    'prop(num_top_production_companies)',
    'prop(num_top_artists)']),
  ('Standard', StandardScaler(), ['runtime', 'release_date_timestamp']),
  ('OneHot_split_genres',
   CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x00000234940BAA60>),
   'genres'),
  ('OneHot_split_credits',
   CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x00000234940BA670>),
   'credits'),
  ('OneHot_split_production_companies',
   CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x0000023494360BE0>),
   'production_companies'),
  ('OneHot_split_keywords',
   CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x0000023494360040>),
   'keywords')],
 'ColumnTransformer__verbose': False,
 'ColumnTransformer__verbose_feature_names_out': True,
 'ColumnTransformer__OneHot': OneHotEncoder(handle_unknown='ignore'),
 'ColumnTransformer__MinMax': MinMaxScaler(),
 'ColumnTransformer__Standard': StandardScaler(),
 'ColumnTransformer__OneHot_split_genres': CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x00000234940BAA60>),
 'ColumnTransformer__OneHot_split_credits': CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x00000234940BA670>),
 'ColumnTransformer__OneHot_split_production_companies': CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x0000023494360BE0>),
 'ColumnTransformer__OneHot_split_keywords': CountVectorizer(tokenizer=<__main__.SplitTokenizer object at 0x0000023494360040>),
 'ColumnTransformer__OneHot__categories': 'auto',
 'ColumnTransformer__OneHot__drop': None,
 'ColumnTransformer__OneHot__dtype': numpy.float64,
 'ColumnTransformer__OneHot__handle_unknown': 'ignore',
 'ColumnTransformer__OneHot__sparse': True,
 'ColumnTransformer__MinMax__clip': False,
 'ColumnTransformer__MinMax__copy': True,
 'ColumnTransformer__MinMax__feature_range': (0, 1),
 'ColumnTransformer__Standard__copy': True,
 'ColumnTransformer__Standard__with_mean': True,
 'ColumnTransformer__Standard__with_std': True,
 'ColumnTransformer__OneHot_split_genres__analyzer': 'word',
 'ColumnTransformer__OneHot_split_genres__binary': False,
 'ColumnTransformer__OneHot_split_genres__decode_error': 'strict',
 'ColumnTransformer__OneHot_split_genres__dtype': numpy.int64,
 'ColumnTransformer__OneHot_split_genres__encoding': 'utf-8',
 'ColumnTransformer__OneHot_split_genres__input': 'content',
 'ColumnTransformer__OneHot_split_genres__lowercase': True,
 'ColumnTransformer__OneHot_split_genres__max_df': 1.0,
 'ColumnTransformer__OneHot_split_genres__max_features': None,
 'ColumnTransformer__OneHot_split_genres__min_df': 1,
 'ColumnTransformer__OneHot_split_genres__ngram_range': (1, 1),
 'ColumnTransformer__OneHot_split_genres__preprocessor': None,
 'ColumnTransformer__OneHot_split_genres__stop_words': None,
 'ColumnTransformer__OneHot_split_genres__strip_accents': None,
 'ColumnTransformer__OneHot_split_genres__token_pattern': '(?u)\\b\\w\\w+\\b',
 'ColumnTransformer__OneHot_split_genres__tokenizer': <__main__.SplitTokenizer at 0x234940baa60>,
 'ColumnTransformer__OneHot_split_genres__vocabulary': None,
 'ColumnTransformer__OneHot_split_credits__analyzer': 'word',
 'ColumnTransformer__OneHot_split_credits__binary': False,
 'ColumnTransformer__OneHot_split_credits__decode_error': 'strict',
 'ColumnTransformer__OneHot_split_credits__dtype': numpy.int64,
 'ColumnTransformer__OneHot_split_credits__encoding': 'utf-8',
 'ColumnTransformer__OneHot_split_credits__input': 'content',
 'ColumnTransformer__OneHot_split_credits__lowercase': True,
 'ColumnTransformer__OneHot_split_credits__max_df': 1.0,
 'ColumnTransformer__OneHot_split_credits__max_features': None,
 'ColumnTransformer__OneHot_split_credits__min_df': 1,
 'ColumnTransformer__OneHot_split_credits__ngram_range': (1, 1),
 'ColumnTransformer__OneHot_split_credits__preprocessor': None,
 'ColumnTransformer__OneHot_split_credits__stop_words': None,
 'ColumnTransformer__OneHot_split_credits__strip_accents': None,
 'ColumnTransformer__OneHot_split_credits__token_pattern': '(?u)\\b\\w\\w+\\b',
 'ColumnTransformer__OneHot_split_credits__tokenizer': <__main__.SplitTokenizer at 0x234940ba670>,
 'ColumnTransformer__OneHot_split_credits__vocabulary': None,
 'ColumnTransformer__OneHot_split_production_companies__analyzer': 'word',
 'ColumnTransformer__OneHot_split_production_companies__binary': False,
 'ColumnTransformer__OneHot_split_production_companies__decode_error': 'strict',
 'ColumnTransformer__OneHot_split_production_companies__dtype': numpy.int64,
 'ColumnTransformer__OneHot_split_production_companies__encoding': 'utf-8',
 'ColumnTransformer__OneHot_split_production_companies__input': 'content',
 'ColumnTransformer__OneHot_split_production_companies__lowercase': True,
 'ColumnTransformer__OneHot_split_production_companies__max_df': 1.0,
 'ColumnTransformer__OneHot_split_production_companies__max_features': None,
 'ColumnTransformer__OneHot_split_production_companies__min_df': 1,
 'ColumnTransformer__OneHot_split_production_companies__ngram_range': (1, 1),
 'ColumnTransformer__OneHot_split_production_companies__preprocessor': None,
 'ColumnTransformer__OneHot_split_production_companies__stop_words': None,
 'ColumnTransformer__OneHot_split_production_companies__strip_accents': None,
 'ColumnTransformer__OneHot_split_production_companies__token_pattern': '(?u)\\b\\w\\w+\\b',
 'ColumnTransformer__OneHot_split_production_companies__tokenizer': <__main__.SplitTokenizer at 0x23494360be0>,
 'ColumnTransformer__OneHot_split_production_companies__vocabulary': None,
 'ColumnTransformer__OneHot_split_keywords__analyzer': 'word',
 'ColumnTransformer__OneHot_split_keywords__binary': False,
 'ColumnTransformer__OneHot_split_keywords__decode_error': 'strict',
 'ColumnTransformer__OneHot_split_keywords__dtype': numpy.int64,
 'ColumnTransformer__OneHot_split_keywords__encoding': 'utf-8',
 'ColumnTransformer__OneHot_split_keywords__input': 'content',
 'ColumnTransformer__OneHot_split_keywords__lowercase': True,
 'ColumnTransformer__OneHot_split_keywords__max_df': 1.0,
 'ColumnTransformer__OneHot_split_keywords__max_features': None,
 'ColumnTransformer__OneHot_split_keywords__min_df': 1,
 'ColumnTransformer__OneHot_split_keywords__ngram_range': (1, 1),
 'ColumnTransformer__OneHot_split_keywords__preprocessor': None,
 'ColumnTransformer__OneHot_split_keywords__stop_words': None,
 'ColumnTransformer__OneHot_split_keywords__strip_accents': None,
 'ColumnTransformer__OneHot_split_keywords__token_pattern': '(?u)\\b\\w\\w+\\b',
 'ColumnTransformer__OneHot_split_keywords__tokenizer': <__main__.SplitTokenizer at 0x23494360040>,
 'ColumnTransformer__OneHot_split_keywords__vocabulary': None,
 'selector__percentile': 95,
 'selector__score_func': <function sklearn.feature_selection._univariate_selection.f_classif(X, y)>,
 'clf__C': 0.001,
 'clf__class_weight': 'balanced',
 'clf__dual': True,
 'clf__fit_intercept': True,
 'clf__intercept_scaling': 1,
 'clf__loss': 'squared_hinge',
 'clf__max_iter': 1000,
 'clf__multi_class': 'ovr',
 'clf__penalty': 'l2',
 'clf__random_state': 42,
 'clf__tol': 0.0001,
 'clf__verbose': 0}
